### PR TITLE
build(docker): upgrade base image to ubuntu 24.04

### PR DESCRIPTION
**Pull request by wirespecter**
Closes https://github.com/reanahub/reana/issues/808
Old PR: https://github.com/reanahub/reana-workflow-engine-snakemake/pull/98

### CODE REVIEW

**Comment by mdonadoni** [comment]
It seems like we have some issues after the upgrade.

In particular, this workflow fails:

Snakefile
```snakefile
```

reana.yaml
```yaml
```

File structure (input files are empty)
```console
$ tree
.
├── inputs
│   ├── a.txt
│   ├── b.txt
│   └── c.txt
├── reana.yaml
└── Snakefile

2 directories, 5 files
```

Error reported by REANA
```
2024-07-25 15:32:02,164 | snakemake.logging | MainThread | CRITICAL | File path ' inputs/ a.txt ' ends with whitespace. This is likely unintended. It can also lead to inconsistent results of the file-matching approach used by Snakemake.
2024-07-25 15:32:02,168 | snakemake.logging | MainThread | ERROR | MissingInputException in rule merge in file /var/reana/users/00000000-0000-0000-0000-000000000000/workflows/fbd5cc04-f1be-4078-b1db-4b55809d3b0b/Snakefile, line 11:
Missing input files for rule merge:
    output: results/merged.txt
    affected files:
         inputs/ a.txt 
         inputs/ c.txt 
         inputs/ b.txt 
```
For some reason, the files have whitespaces in their filename, we can have a look together

File: Dockerfile
Comment (mdonadoni) [inline Dockerfile:1]: There is one comment in this file which mentions Ubuntu 20.04. Let's also update that!

File: Dockerfile [hunk 1]
```diff
@@ -35,14 +33,14 @@ RUN apt-get update -y && \
...
+        libauthen-krb5-simple-perl \
         libkrb5-dev \
         libssl-dev \
         make \
         pkg-config \
-        python3-dev \
+        python3.12-dev \
         python3-pip \
-        python3.8 \
+        python3.12 \
```
Comment (tiborsimko) [inline Dockerfile:43]: BTW the Snakemake py312 f-string related fix in https://github.com/reanahub/reana-commons/pull/465 now requires `git` executable to be present in the image in order to install our Snakemake patches, so you can proactively add `git` to the list of requirements here.

Comment (mdonadoni) [reply Dockerfile:43]: This is the diff of the changes I was using:
```diff
diff --git a/Dockerfile b/Dockerfile
index df7ca56..694f258 100644
--- a/Dockerfile
+++ b/Dockerfile
@@ -27,6 +27,7 @@ RUN apt-get update -y && \
         curl \
         g++ \
         gcc \
+        git \
         gnupg2 \
         graphviz \
         graphviz-dev \
@@ -64,6 +65,7 @@ RUN apt-get update -y && \
         cmake \
         g++ \
         gcc \
+        git \
         graphviz-dev \
         libssl-dev \
         make \
@@ -91,13 +93,22 @@ RUN if [ "${DEBUG}" -gt 0 ]; then pip install --no-cache-dir -e ".[debug,xrootd]
 
 # Are we building with locally-checked-out shared modules?
 # hadolint ignore=DL3013
-RUN if test -e modules/reana-commons; then \
+RUN apt-get update -y && \
+    apt-get install -y --no-install-recommends \
+      git && \
+    if test -e modules/reana-commons; then \
       if [ "${DEBUG}" -gt 0 ]; then \
         pip install --no-cache-dir -e "modules/reana-commons[snakemake_reports]" --upgrade; \
       else \
         pip install --no-cache-dir "modules/reana-commons[snakemake_reports]" --upgrade; \
       fi \
-    fi
+    fi && \
+    apt-get remove -y \
+        git && \
+    apt-get autoremove -y && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/*
+
```

**Review by tiborsimko — APPROVED** [review]

**Pull request merged**